etc/sync-function-template.js

```diff
@@ -99,31 +99,39 @@ function(doc, oldDoc) {
   function validateDoc(doc, oldDoc, docDefinition, docType) {
     var validationErrors = [ ];
 
-    if (!(docDefinition.allowAttachments) && doc._attachments) {
-      for (var attachment in doc._attachments) {
-        validationErrors.push('document type does not support attachments');
-
-        break;
-      }
+    if (docDefinition.immutable && oldDoc && !(oldDoc._deleted)) {
+      validationErrors.push('documents of this type cannot be replaced or deleted');
     }
 
-    var itemStack = [
-      {
-        itemValue: doc,
-        oldItemValue: oldDoc,
-        itemName: null
+    // Only validate the document's contents if it's being created or replaced. But there's no need if it's being deleted.
+    if (!doc._deleted) {
+      if (!(docDefinition.allowAttachments) && doc._attachments) {
+        for (var attachment in doc._attachments) {
+          validationErrors.push('document type does not support attachments');
+
+          break;
+        }
       }
-    ];
 
-    // Execute each of the document's property validators
-    validateProperties(doc, oldDoc, docDefinition.propertyValidators, itemStack, validationErrors, true);
+      var itemStack = [
+        {
+          itemValue: doc,
+          oldItemValue: oldDoc,
+          itemName: null
+        }
+      ];
+
+      // Execute each of the document's property validators while ignoring these whitelisted properties at the root level
+      var whitelistedProperties = [ '_id', '_rev', '_deleted', '_revisions', '_attachments' ];
+      validateProperties(doc, oldDoc, docDefinition.propertyValidators, itemStack, validationErrors, whitelistedProperties);
```

Comment:
Passing in the whitelisted property names as an array is a more flexible option than hardcoding them into the `validateProperties` function.